File: lib/src/main/java/com/pigdogbay/lib/utils/ObservableProperty.java
package com.pigdogbay.lib.utils;

import java.util.ArrayList;
import java.util.List;

/**
 * See com.pigdogbay.lib.patterns.ObservableProperty for specifying a sender
 * Here sender is an instance of this class
 * @param <T> Property Type
 */
public class ObservableProperty<T>
{
	public interface PropertyChangedObserver<T>
	{
		void update(ObservableProperty<T> sender, T update);
	}
	private T value;
	private List<PropertyChangedObserver<T>> observers;

	public ObservableProperty(T value)
	{
		this.value = value;
		observers = new ArrayList<>();
	}
	
	public synchronized T getValue()
	{
		return value;
	}
	public void setValue(T newValue)
	{
		this.value = newValue;
		for (PropertyChangedObserver<T> listener : observers)
		{
			listener.update(this, newValue);
		}
	}
	public void setValueWithoutNotification(T newValue)
	{
		this.value = newValue;
	}
	
	public synchronized void addObserver(PropertyChangedObserver<T> observer)
	{
		observers.add(observer);
	}
	public synchronized void removeObserver(PropertyChangedObserver<T> observer)
	{
		observers.remove(observer);
	}
}

File: lib/src/main/java/com/pigdogbay/lib/patterns/ObservableProperty.kt
package com.pigdogbay.lib.patterns

import java.util.ArrayList


/**
 * Observers must implement this interface to receive updates when the property changes
 */
interface PropertyChangedObserver<T, U> {
    fun update(sender: T, update: U)
}

/**
 * Observable property that allows you to specify the sender for updates
 * Updates are sent when the property value changes
 * See also utils.ObservableProperty
 */
class ObservableProperty<T, U>(private val sender: T, initialValue: U) {

    private var backingValue = initialValue
    private val observers: MutableList<PropertyChangedObserver<T, U>> = ArrayList()

    var value
        @Synchronized get() = backingValue
        @Synchronized set(newValue) {
            if (backingValue!=newValue) {
                this.backingValue = newValue
                for (listener in observers) {
                    listener.update(sender, newValue)
                }
            }
        }

    fun setValueWithoutNotification(newValue: U) {
        this.value = newValue
    }

    @Synchronized
    fun addObserver(observer: PropertyChangedObserver<T, U>) {
        observers.add(observer)
    }

    @Synchronized
    fun removeObserver(observer: PropertyChangedObserver<T, U>) {
        observers.remove(observer)
    }

    @Synchronized
    fun removeAllObservers(){
        observers.clear()
    }
}
